### PR TITLE
Improved keyboard interrupt handling backend

### PR DESCRIPTION
**Pull request by hoodmane**
This is a replacement for pyodide-interrupts.
It fixes two quirks in pyodide-interrupts: It interacts correctly with sys.settrace, and it uses signals correctly.
As a result, if you use `signal.signal(SIGINT, some_handler)` then `some_handler` will be called when an interrupt is triggered.

In order to interact correctly with `sys.settrace`, I have to monkey patch `sys.settrace`. In particular, I copied the entire definition of `sys.settrace` into my code and added one extra line to call my extra interrupt handling. This has the disadvantage that it pins a specific version of `sys.settrace` from Python 3.8.2.

If people prefer I could keep this as a separate package rather than integrating it into Pyodide core.

### CODE REVIEW

**Comment by hoodmane** [comment]
The monkey patch of `sys.settrace` was really bugging me. When I was looking into the inner workings of the `signal` package, I found `Py_AddPendingCall` which sounded like what I wanted. And indeed, `Py_AddPendingCall` can be used for this purpose avoiding all monkey patching. Currently one of my tests is failing but I'm confident this approach can be used and it's a significant improvement.

**Comment by hoodmane** [comment]
Looks like the failing test passes on CI, I think it's something screwed up about the state of my local copy and doing a fresh build will make it go away.

File: src/type_conversion/interrupts.c [hunk 1]
```diff
@@ -0,0 +1,228 @@
...
+  //   goto fail;
+  // }
+
+  // func = PyCFunction_NewEx(&settrace_methoddef, (PyObject*)module, name);
+  // if (func == NULL) {
+  //   goto fail;
+  // }
+  // if (PyObject_SetAttrString(module, "settrace", func)) {
+  //   goto fail;
+  // }
```
Comment (rth) [inline src/type_conversion/interrupts.c:181]: Commented code should be removed.

File: src/pyodide.py [hunk 2]
```diff
@@ -7,6 +7,31 @@
...
+def interrupt_buffer(callback):
+    """
+    Periodically polls ``callback``. If ``callback`` returns a nonzero value, triggers a ``SIGINT`` signal.
+    By default, the signal handler for ``SIGINT`` raises a ``KeyboardException``, but using the ``signals`` package this can be changed.
+
+    If ``callback`` returns a value that cannot be interpreted as an integer or if an exception is triggered inside of ``callback``
+    then the exception is allowed to propogate but the interrupt buffer is set to ``None``.
+
+    Args:
+        callback -- a zero argument function which returns an int. If it returns a nonzero value, triggers ``SIGINT``.
```
Comment (rth) [inline src/pyodide.py:27]: We use numpy docstring format https://numpydoc.readthedocs.io/en/latest/format.html#docstring-standard 

So it needs a 1 line summary, and then something like
```
Parameters
----------
callback : callable
    a zero ...
```

File: src/type_conversion/interrupts.c [hunk 3]
```diff
@@ -0,0 +1,228 @@
...
+             "can be changed."
+             "If ``callback`` returns a value that cannot be interpreted as an "
+             "integer or if an exception is triggered inside of ``callback``"
+             "then the exception is allowed to propagate but the interrupt "
+             "buffer is set to ``None``."
+             "If called with ``None``, interrupt polling is turned off."
+             "\n"
+             "Args:\n"
+             "   callback -- a zero argument function which returns an int. If "
+             "it returns a nonzero value, triggers ``SIGINT``.\n");
```
Comment (rth) [inline src/type_conversion/interrupts.c:75]: Same comment about docstring format as above.

**Review by rth — COMMENTED** [review]
Thanks @hoodmane ! You would need to fix merge conflicts following #909 . Maybe we could put it under `src/pyodide-py/pyodide/interrupt.py`

> If people prefer I could keep this as a separate package rather than integrating it into Pyodide core.

I'm not opposed to merging this into pyodide code. However I won't have the bandwidth to review this in detail before the release #831 . So unless @dalcde looks into it, it might be easier to update it as a separate package for now.

Also please update the changelog.

**Comment by dalcde** [comment]
Two comments:
1. Can we benchmark the performance impact of using this? I recall that the previous `settrace` method has essentially undetectable performance impact. Is the same true for `Py_AddPendingCall`?
2. Can we add interrupt functionality to `webworker.js`? Hopefully we can integration test the interrupt functionality this way (e.g. with a sleep loop). This will let us close #676

**Comment by hoodmane** [comment]
> Can we benchmark the performance impact of using this?

We should do that before merging this.

`Py_AddPendingCall` acquires a lock (always open when there is only one thread), pushes the callback onto a stack, and then does a relaxed atomic store into a signal that there are pending callbacks so that is effectively free.

Oh uhh and then on each step through the `ceval` main loop it calls my callback 32 times because `NPENDINGCALLS` is `32` and the callback recursively sets itself. For most of these 32 calls my callback just decrements a counter and returns, the cost of this is probably (?) still negligible. Annoyingly there is no clear way to defer the next call to after the next python instruction is executed.

Currently I know of four possible entrypoints to add code to the main loop:
`PyEval_SetTrace`, `PyEval_SetProfile`, setting a custom `PyInterpreterState.eval_frame` and `Py_AddPendingCall`. 

Of these four options, the only one that mixes well with all possible other code is `Py_AddPendingCall`. However, `eval_frame` is a good option because it is not exposed to the Python user directly like `sys.settrace` and `sys.setprofile` are. Custom `eval_frame` is currently only used for JIT, and I don't expect us to support JITs for Pyodide any time soon...

**Comment by hoodmane** [comment]
Oh `PyInterpreterState.eval_frame` doesn't work because it only runs once per stack frame which isn't enough: we want keyboard interrupts to work even if the interpreter is in an infinite loop with no function calls in it, such as `while True: pass`

**Comment by hoodmane** [comment]
> I'm not opposed to merging this into pyodide code. However I won't have the bandwidth to review this in detail before the release #831.

This is more of a fine tuning details PR, the current `pyodide-interrupts` code works for practical purposes so there is no rush on including it until we get it right.

**Comment by hoodmane** [comment]
I am now thinking that the best approach to this might be to add a patch to `ceval.c` to expose an extra callback. @dalcde @rth Would you guys object to that? You have any thoughts?

**Comment by hoodmane** [comment]
Closed in favor of #1148.

**Pull request closed without merging**